File: State/cpp-source/states/Yellow.cpp
//
// Created by Pavel Akhtyamov on 2019-04-23.
//

#include "Yellow.h"
#include "../TrafficLight.h"
#include "Red.h"
#include "Green.h"


void Yellow::Switch(TrafficLight *light) {
  if (light->GetDirection() == Direction::TO_GREEN) {
    light->SetState(std::make_unique<Green>());
  } else {
    light->SetState(std::make_unique<Red>());
  }
}

std::string Yellow::GetStringState() {
  return "Yellow";
}

File: JavaSource/src/main/java/ru/akhcheck/patterns/state/states/Yellow.java
package ru.akhcheck.patterns.state.states;

import ru.akhcheck.patterns.state.Direction;
import ru.akhcheck.patterns.state.TrafficLight;

/**
 * Жёлтый единственный, кому нужно направление.
 *
 * <p>Из красного он ведёт в зелёный, из зелёного в красный, и без хранимого
 * направления отличить эти два случая нельзя.
 */
public class Yellow implements State {
    @Override
    public void switchState(TrafficLight light) {
        if (light.getDirection() == Direction.TO_GREEN) {
            light.setState(new Green());
            return;
        }
        light.setState(new Red());
    }

    @Override
    public String getStringState() {
        return "Yellow";
    }
}
